### PR TITLE
New API Plugin - Atomic metadata operations (master)

### PR DESCRIPTION
**Pull request by korydraughn**
Requires irods/externals#45

### CODE REVIEW

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+                    {"attribute", "this_was_set_atomically"},
+                    {"value", "it worked!"}
+                }
+            })}
+        }.dump();
+
+        char* json_error_string{};
+
+        REQUIRE(rc_atomic_apply_metadata_operations(&_conn, json_input.c_str(), &json_error_string) == 0);
+        REQUIRE(json_error_string == "{}"s);
```
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

File: plugins/api/src/atomic_apply_metadata_operations.cpp [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+    }
+
+    auto to_bytes_buffer(const std::string& _s) -> bytesBuf_t*
+    {
+        const auto buf_size = _s.length() + 1;
+
+        char* buf = new char[buf_size]{};
+        std::strncpy(buf, _s.c_str(), _s.length());
+
+        bytesBuf_t* bbp = new bytesBuf_t{};
```
Comment (trel) [inline plugins/api/src/atomic_apply_metadata_operations.cpp]: Codacy Issue found: [Uninitialized variable: bbp](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp (hunk 1 above)
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

File: plugins/api/src/atomic_apply_metadata_operations.cpp [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+            }
+            else if (op_code == "remove") {
+                if (const auto meta_id = get_meta_id(_db_conn, md); meta_id > -1) {
+                    detach_metadata_from_object(_db_conn, _object_id, meta_id);
+                }
+            }
+            else {
+                // clang-format off
+                log::api::error({{"log_message", "Invalid metadata operation"},
+                                 {"metadata_operation", _op}});
```
Comment (korydraughn) [inline plugins/api/src/atomic_apply_metadata_operations.cpp:375]: Shouldn't _op be a string?

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp (hunk 1 above)
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp:474]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp:474]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp:474]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

**Comment by korydraughn** [comment]
CI core tests passed on Ubuntu 16.
Did not test Centos 7.

**Comment by korydraughn** [comment]
Plugin was not being tested. Fixing and rerunning tests.

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
```
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp:475]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5179166)

**Comment by korydraughn** [comment]
Is this an admin-only function?

**Comment by trel** [comment]
i don't think so - any user should be allowed to perform atomic metadata operations - there shouldn't be any new permissions issues here...

File: lib/api/include/atomic_apply_metadata_operations.h [hunk 5]
```diff
@@ -1,57 +1,76 @@
...
- *
- * \param[in] _comm - A pointer to a rcComm_t.
- * \param[in] _json_input - A JSON string containing the batch of metadata updates.
- * \param[out] _json_output - A JSON string containing the error information on failure.
- *
- * \return integer
- * \retval 0 on success, non-zero on failure.
- **/
-int rc_atomic_apply_metadata_operations(rcComm_t* _comm, const char* _json_input, char** _json_output);
+/// Executes a list of metadata operations on a single object atomically.
```
Comment (trel) [inline lib/api/include/atomic_apply_metadata_operations.h:12]: why three forward slashes?   is that better for some reason?

Comment (korydraughn) [reply lib/api/include/atomic_apply_metadata_operations.h:12]: Allows doxygen to see it and that appears to be the style used to comment new things in iRODS.

File: plugins/api/src/atomic_apply_metadata_operations.cpp [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+                                           " a.user_id = (select user_id from R_USER_MAIN where user_name = '{}') and"
+                                           " a.object_id = '{}'", _comm.clientUser.userName, _object_id);
+
+            if (auto row = execute(_db_conn, query); row.next()) {
+                constexpr int access_modify_object = 1120;
+                return row.get<int>(0) >= access_modify_object;
+            }
+        }
+        else if (_entity_type == "user" || _entity_type == "resource") {
+            return irods::is_privileged_client(_comm);
```
Comment (korydraughn) [inline plugins/api/src/atomic_apply_metadata_operations.cpp:223]: Should applying metadata to users and resources require admin-level privileges? Can rodsusers apply metadata to users and resources now?

**Comment by korydraughn** [comment]
CI core tests passed on Ubuntu 16.
Did not test Centos 7.
Verified that plugin tests ran.

File: plugins/api/src/atomic_apply_metadata_operations.cpp (hunk 6 above)
Comment (korydraughn) [reply plugins/api/src/atomic_apply_metadata_operations.cpp:223]: Based on `imeta`'s docs, the answer is users must be admins to set metadata on users and resources.

**Comment by korydraughn** [comment]
Added #

File: plugins/api/src/atomic_apply_metadata_operations.cpp (hunk 6 above)
Comment (trel) [reply plugins/api/src/atomic_apply_metadata_operations.cpp:223]: yes, confirmed.

**Pull request merged**